### PR TITLE
Flashbangfix + Shot fix

### PR DESCRIPTION
**Pull request by Undid-Iridium**

### CODE REVIEW

**Comment by Undid-Iridium** [comment]
> Will need to finish Transpiler disable 

What.

**Review by Thundermaker300 — APPROVED** [review]

File: Exiled.Events/Patches/Events/Map/BreakingScp2176.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ namespace Exiled.Events.Patches.Events.Map
     ///     Patches <see cref="Scp2176Projectile.ServerShatter" />.
     ///     Supplements the <see cref="Map.ExplodingGrenade" /> event.
     /// </summary>
-    [HarmonyPatch(typeof(Scp2176Projectile), nameof(Scp2176Projectile.ServerShatter))]
+    // [HarmonyPatch(typeof(Scp2176Projectile), nameof(Scp2176Projectile.ServerShatter))]
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Map/BreakingScp2176.cs:32]: can't we keep this not comment ?

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 2]
```diff
@@ -36,108 +34,56 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                    new(OpCodes.Callvirt, Method(typeof(ListPool<ReferenceHub>), nameof(ListPool<ReferenceHub>.Pool.Return))),
-
-                    // return;
-                    new CodeInstruction(OpCodes.Ret).WithLabels(returnLabel),
+                    // FlashbangGrenade
+                    new CodeInstruction(OpCodes.Ldarg_0),
+
+                    // Returns DoctorSenseEventArgs
+                    new(OpCodes.Call, Method(typeof(ExplodingFlashGrenade), nameof(ProcessEvent))),
+                    new(OpCodes.Br, returnLabel),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:52]:
```suggestion
                    new(OpCodes.Br_S, returnLabel),
```

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 3]
```diff
@@ -36,108 +34,56 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
+        private static void ProcessEvent(FlashbangGrenade instance)
         {
-            foreach (Player player in players)
+            ExplodingGrenadeEventArgs explodingGrenadeEvent = new ExplodingGrenadeEventArgs(Player.Get(instance.PreviousOwner.Hub), instance);
+            Handlers.Map.OnExplodingGrenade(explodingGrenadeEvent);
+
+            if (!explodingGrenadeEvent.IsAllowed)
+            {
+                return;
+            }
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:71]:
```suggestion
            if (!explodingGrenadeEvent.IsAllowed)
                return;
```

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 4]
```diff
@@ -36,108 +34,56 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
         }
 
-        private static List<Player> ConvertHubs(List<ReferenceHub> hubs) => hubs.Select(Player.Get).ToList();
-
-        private static void ProcessPlayers(FlashbangGrenade grenade, List<Player> players)
+        private static void ProcessEvent(FlashbangGrenade instance)
         {
-            foreach (Player player in players)
+            ExplodingGrenadeEventArgs explodingGrenadeEvent = new ExplodingGrenadeEventArgs(Player.Get(instance.PreviousOwner.Hub), instance);
+            Handlers.Map.OnExplodingGrenade(explodingGrenadeEvent);
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:66]:
```suggestion
            ExplodingGrenadeEventArgs explodingGrenadeEvent = new ExplodingGrenadeEventArgs(Player.Get(instance.PreviousOwner.Hub), instance);

            Handlers.Map.OnExplodingGrenade(explodingGrenadeEvent);
```

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 5]
```diff
@@ -36,108 +34,56 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                    grenade.ProcessPlayer(player.ReferenceHub);
+                if (!ExiledEvents.Instance.Config.CanFlashbangsAffectThrower && (explodingGrenadeEvent.Player == player))
+                {
+                    continue;
+                }
+
+                if (HitboxIdentity.CheckFriendlyFire(explodingGrenadeEvent.Player.ReferenceHub, player.ReferenceHub))
+                {
+                    instance.ProcessPlayer(player.ReferenceHub);
+                }
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:83]:
```suggestion
                if (!ExiledEvents.Instance.Config.CanFlashbangsAffectThrower && (explodingGrenadeEvent.Player == player))
                    continue;

                if (HitboxIdentity.CheckFriendlyFire(explodingGrenadeEvent.Player.ReferenceHub, player.ReferenceHub))
                    instance.ProcessPlayer(player.ReferenceHub);
```

File: Exiled.Events/Patches/Events/Player/Shot.cs [hunk 6]
```diff
@@ -94,64 +82,58 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
+        private static bool ProcessShot(ReferenceHub player, RaycastHit hit, IDestructible destructible,  ref float damage)
+        {
+            ShotEventArgs shotEvent = new ShotEventArgs(Player.Get(player), hit, destructible, damage);
+            Handlers.Player.OnShot(shotEvent);
+            if (shotEvent.CanHurt)
+            {
+                damage = shotEvent.Damage;
             }
+
+            return shotEvent.CanHurt;
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Shot.cs:136]:
```suggestion
        private static bool ProcessShot(ReferenceHub player, RaycastHit hit, IDestructible destructible, ref float damage)
        {
            ShotEventArgs shotEvent = new ShotEventArgs(Player.Get(player), hit, destructible, damage);
            
            Handlers.Player.OnShot(shotEvent);
            
            if (shotEvent.CanHurt)
                damage = shotEvent.Damage;

            return shotEvent.CanHurt;
```

File: Exiled.Events/Patches/Events/Player/Shot.cs [hunk 7]
```diff
@@ -94,64 +82,58 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                        new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnShot))),
+                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot))),
 
-                        // if (!ev.CanHurt)
-                        //    return;
-                        new(OpCodes.Callvirt, PropertyGetter(typeof(ShotEventArgs), nameof(ShotEventArgs.CanHurt))),
-                        new(OpCodes.Brfalse, returnLabel),
+                    // if (!ev.CanHurt)
+                    //    return;
+                    new(OpCodes.Brfalse, returnLabel),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Shot.cs:116]:
```suggestion
                    new(OpCodes.Brfalse_S, returnLabel),
```

File: Exiled.Events/Patches/Events/Player/Shot.cs [hunk 8]
```diff
@@ -94,64 +82,58 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                        new(OpCodes.Newobj, GetDeclaredConstructors(typeof(ShotEventArgs))[0]),
-                        new(OpCodes.Dup),
-                        new(OpCodes.Dup),
-                        new(OpCodes.Stloc, ev.LocalIndex),
+                    // damage
+                    new(OpCodes.Ldloca_S, 4),
 
-                        // Handlers.Player.OnShot(ev)
-                        new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnShot))),
+                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot))),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Shot.cs:112]:
```suggestion
                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot), new[] { typeof(ReferenceHub), typeof(RaycastHit), typeof(IDestructible),  typeof(float).MakeByRefType()})),
```

File: Exiled.Events/Patches/Events/Player/Shot.cs [hunk 9]
```diff
@@ -59,26 +60,13 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                    new(OpCodes.Stloc, ev.LocalIndex),
-
-                    // Handlers.Player.OnShot(ev)
-                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnShot))),
+                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot))),
 
                     // if (!ev.CanHurt)
                     //    return;
-                    new(OpCodes.Callvirt, PropertyGetter(typeof(ShotEventArgs), nameof(ShotEventArgs.CanHurt))),
                     new(OpCodes.Brfalse, returnLabel),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Shot.cs:69]:
```suggestion
                    new(OpCodes.Brfalse_S, returnLabel),
```

File: Exiled.Events/Patches/Events/Player/Shot.cs [hunk 10]
```diff
@@ -59,26 +60,13 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
 
-                    // ShotEventArgs ev = new(Player, RaycastHit, IDestructible, float)
-                    new(OpCodes.Newobj, GetDeclaredConstructors(typeof(ShotEventArgs))[0]),
-                    new(OpCodes.Dup),
-                    new(OpCodes.Dup),
-                    new(OpCodes.Stloc, ev.LocalIndex),
-
-                    // Handlers.Player.OnShot(ev)
-                    new(OpCodes.Call, Method(typeof(Handlers.Player), nameof(Handlers.Player.OnShot))),
+                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot))),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Shot.cs:65]:
```suggestion
                    new(OpCodes.Call, Method(typeof(Shot), nameof(ProcessShot), new[] { typeof(ReferenceHub), typeof(RaycastHit), typeof(IDestructible),  typeof(float).MakeByRefType()})),
```

File: Exiled.Events/Patches/Events/Map/BreakingScp2176.cs (hunk 1 above)
Comment (Undid-Iridium) [reply Exiled.Events/Patches/Events/Map/BreakingScp2176.cs:32]: Do as IOPietro did and I'll commit it

**Review by iopietro — APPROVED** [review]

**Review by louis1706 — APPROVED** [review]

**Review by iopietro — CHANGES_REQUESTED** [review]
Build errors.

**Comment by Undid-Iridium** [comment]
The fact there is not just a simple ruleset for a formatter is a really sad part of Exiled, and wastes a lot of time. Action on Save, CSharpier, Prettier, dotnet-format. Several options with configurations/rulesets that could be defined but ¯\_(ツ)_/¯
image

It would be a lot simpler than the current waste of time, and honestly it's almost all opinionated arbitrary preferences, such as no braces for single if conditions (which is normally NOT recommended). You dev-bois should really figure out what rules you want, and then make it a requirement to have than stylecop whose purpose is just to complain instead of fix.    

For example, brace spacing 

image

image
image

image

Anyway, that's my ted talk rant. 

[editorconfig TXT](https://github.com/Exiled-Team/EXILED/files/10505981/editorconfig.TXT)

**Comment by iopietro** [comment]
@Undid-Iridium

I mean, apart from StyleCop rules the only three formatting rules that we are following right now are: 
- Do not clog up your code and space it as much as possible to make it more understandable;
- Do not use braces in single-line ifs;
- Try to keep inline methods in one line when possible.

That's it, if you've some other issues about the formatter you can make a PR about the `.editorconfig` file to reflect something that you want to add/change.

**Review by iopietro — CHANGES_REQUESTED** [review]
Don't use `var`.

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 11]
```diff
@@ -36,108 +34,51 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
+            if (!explodingGrenadeEvent.IsAllowed)
+                return;
+
+            foreach (Player player in explodingGrenadeEvent.TargetsToAffect)
             {
-                if (ExiledEvents.Instance.Config.CanFlashbangsAffectThrower && (Player.Get(grenade.PreviousOwner.Hub) == player))
-                    grenade.ProcessPlayer(player.ReferenceHub);
-                else if (HitboxIdentity.CheckFriendlyFire(grenade.PreviousOwner.Hub, player.ReferenceHub))
-                    grenade.ProcessPlayer(player.ReferenceHub);
+                if (!ExiledEvents.Instance.Config.CanFlashbangsAffectThrower && (explodingGrenadeEvent.Player == player))
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:74]:
```suggestion
                if (!ExiledEvents.Instance.Config.CanFlashbangsAffectThrower && explodingGrenadeEvent.Player == player)
```

File: Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs [hunk 12]
```diff
@@ -36,108 +34,51 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-                    new(OpCodes.Call, PropertyGetter(typeof(ListPool<ReferenceHub>), nameof(ListPool<ReferenceHub>.Pool))),
-                    new(OpCodes.Ldloc, list.LocalIndex),
-                    new(OpCodes.Callvirt, Method(typeof(ListPool<ReferenceHub>), nameof(ListPool<ReferenceHub>.Pool.Return))),
-
-                    // return;
-                    new CodeInstruction(OpCodes.Ret).WithLabels(returnLabel),
+                    // FlashbangGrenade
+                    new CodeInstruction(OpCodes.Ldarg_0),
+
+                    // Returns DoctorSenseEventArgs
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ExplodingFlashGrenade.cs:50]:
```suggestion
                    // Processes ExplodingGrenadeEventArgs
```

**Review by iopietro — CHANGES_REQUESTED** [review]

**Pull request merged**